genie-web/src/main/java/com/netflix/genie/web/agent/apis/rpc/v4/endpoints/GRpcAgentFileStreamServiceImpl.java

```diff
@@ -142,333 +151,630 @@ public GRpcAgentFileStreamServiceImpl(
+            this.fileTansferCounter = registry.counter(TRANSFER_COUNTER);
 
-        // Remove observer from the set of transfers waiting to start.
-        final boolean removed = this.pendingTransferObserversSet.remove(fileTransferStreamObserver);
-        if (removed) {
-            log.debug("Removed observer for file stream: {} from 'pending' set", streamId);
+            this.registry = registry;
         }
 
-        // Remove buffer from the set of transfers waiting to start and move it to the 'in progress' map.
-        final StreamBuffer streamBufferFromPending = this.pendingTransferBuffersMap.remove(streamId);
-        if (streamBufferFromPending != null) {
-            log.debug("Moving buffer for file stream: {} from 'pending' to 'in progress'", streamId);
-            this.inProgressTransferBuffersMap.put(streamId, streamBufferFromPending);
-        }
+        private synchronized void requestFile(
+            final String jobId,
+            final String fileTransferId,
+            final String relativePath,
+            final int startOffset,
+            final int endOffset
+        ) throws GenieNotFoundException {
+
+            final ControlStreamObserver controlStreamObserver = this.controlStreamMap.get(jobId);
+            if (controlStreamObserver == null) {
+                throw new GenieNotFoundException("No active stream control stream for job: " + jobId);
+            }
 
-        // Look up the buffer where chunk data is written into
-        final StreamBuffer streamBuffer = this.inProgressTransferBuffersMap.get(streamId);
+            this.fileTansferCounter.increment();
 
-        // Write into it, if the stream is still there
-        if (streamBuffer != null) {
-            log.debug("Writing {} bytes into stream of transfer {}", data.size(), streamId);
-            streamBuffer.write(data);
+            // Send the file request
+            controlStreamObserver.responseObserver.onNext(
+                ServerControlMessage.newBuilder()
+                    .setServerFileRequest(
+                        ServerFileRequestMessage.newBuilder()
+                            .setStreamId(fileTransferId)
+                            .setRelativePath(relativePath)
+                            .setStartOffset(startOffset)
+                            .setEndOffset(endOffset)
+                            .build()
+                    )
+                    .build()
+            );
         }
-    }
 
-    private void handleFileTransferError(
-        final FileTransferStreamObserver fileTransferStreamObserver,
-        @Nullable final String streamId,
-        final Throwable t
-    ) {
-        log.error("Error in file transfer stream: {}: {}", streamId, t.getMessage(), t);
+        private StreamObserver<AgentManifestMessage> handleNewControlStream(
+            final StreamObserver<ServerControlMessage> responseObserver
+        ) {
+            log.info("New agent control stream established");
```

Comment:
Is there a way in the information that is available to log which job this stream is for? Could be helpful for debugging.

Reply:
No, at this point it's just a new "connection". As soon as the first message is sent along in that connection, i can associate the stream to an ongoing transfer and then we do have the rest of the context. But not here.